src/parachain/amm/liquidity-pool/stable.ts

```diff
@@ -0,0 +1,280 @@
+        }
+
+        throw new Error("_getYD: Calculation error.");
+    }
+
+    /**
+     * Sort amounts in same order as `pooledCurrencies`.
+     *
+     * @param amounts Array of monetary
+     * @returns Amounts containing currency amounts at the same index as `this.pooledCurrencies`
+     * @throws When currencies of `amounts` differ from `pooledCurrencies`
+     */
+    private _sortAmounts(amounts: Array<MonetaryAmount<CurrencyExt>>): Array<MonetaryAmount<CurrencyExt>> {
+        if (amounts.length !== this.pooledCurrencies.length) {
+            throw new Error(
+                "StableLiquidityPool: _sortAmounts: Amounts count is different from pooledCurrencies count."
+            );
+        }
+        const sortedAmounts = new Array(amounts.length);
+        for (const amount of amounts) {
+            const indexToSave = this.getTokenIndex(amount.currency);
+
+            const isCurrencyNotFound = indexToSave === -1;
+            const isCurrencyNonunique = sortedAmounts[indexToSave] !== undefined;
+            if (isCurrencyNotFound) {
+                throw new Error(
+                    `StableLiquidityPool: _sortAmounts: Currency ${amount.currency.name} is not part of the pool.`
+                );
+            }
+            if (isCurrencyNonunique) {
+                throw new Error(`StableLiquidityPool: _sortAmounts: Currency ${amount.currency.name} is not unique.`);
+            }
+
+            sortedAmounts[indexToSave] = amount;
+        }
+        return sortedAmounts;
+    }
+
+    public involvesToken(currency: CurrencyExt): boolean {
+        return this.pooledCurrencies.some(({ currency: pooledCurrency }) => isCurrencyEqual(pooledCurrency, currency));
+    }
+
+    public getTokenIndex(currency: CurrencyExt): number {
+        return this.pooledCurrencies.findIndex(({ currency: pooledCurrency }) =>
+            isCurrencyEqual(currency, pooledCurrency)
+        );
+    }
+
+    // TODO: rename to 'currenciesInBaseDenomination'
+    public get xp(): Array<Big> {
+        return this._xp(this.pooledCurrencies);
+    }
+
+    // TODO: rename to something like `calculateLiquidityDeposit`
+    /**
+     *
+     * @param amounts Array of monetary amount for each pooled currency of this pool.
+     * @param deposit
+     * @returns
+     */
+    public calculateTokenAmount(
+        amounts: Array<MonetaryAmount<CurrencyExt>>,
+        deposit: boolean
+    ): MonetaryAmount<CurrencyExt> {
+        const sortedAmounts = this._sortAmounts(amounts);
+
+        const amp = this.A;
+        const D0 = this._getD(this.xp, amp);
+
+        const newBalances = this.pooledCurrencies.map((balance, i) =>
+            deposit ? balance.add(sortedAmounts[i]) : balance.sub(sortedAmounts[i])
+        );
+        const D1 = this._getD(this._xp(newBalances), amp);
+
+        if (this.totalSupply.isZero()) {
+            // TODO: check is D1 in base or atomic denomination??
+            // if not in base divide by decimals first
```

Comment:
If it is in atomic, you can use the lib utility function `newMonetaryAmount()` which accepts atomic (or base)

Reply:
Yes thanks for the reminder!